### DISCUSSION TITLE
Docs: geolocation is confusing (w/vercel) - why is req.geo only available on middlewere

### DISCUSSION THREAD

**Discussion by softmarshmallow** [vercel/next.js#69699] [Help]
### What is the documentation issue?

The doc says 

https://nextjs.org/docs/app/api-reference/functions/next-request#geo
```
Good to know: On [Vercel](https://vercel.com/docs/frameworks/nextjs?utm_source=next-site&utm_medium=docs&utm_campaign=next-website), this value is provided by default. On other platforms, you can use the [X-Forwarded-For](https://developer.mozilla.org/docs/Web/HTTP/Headers/X-Forwarded-For) header to provide the IP address, then use a [third-party service](https://ip-api.com/) to lookup the geographic information.
```

But the req.geo is only available from middleware.js(ts)
and from app router (which ofcource in same project, nextjs, hosted on vercel, no custom configuration) the req.geo will always return `{}` empty object, but when you see headers, all related x-vercel-xxx is available.

which is mentioned on https://vercel.com/guides/geo-ip-headers-geolocation-vercel-functions#using-the-geolocation-helper
```
The geolocation helper returns location information for an incoming request. It takes a request object as input and returns properties such as city, country, latitude, longitude, and region ([Vercel Edge Network region](https://vercel.com/docs/edge-network/regions) that received the request).
```

```
import { geolocation } from '@vercel/functions';
 
export function GET(request: Request) {
  const { city } = geolocation(request);
  return new Response(`<h1>Your location is ${city}</h1>`, {
    headers: { 'content-type': 'text/html' },
  });
}
```

## Question.

- Why is req.geo available only from middleware, if there is both required headers present for middleware.ts and route.ts
- Why does the doc states that `Good to know: On Vercel, this value is provided by default` as it is only available on middleware
- Is it intended to use helper function for route.ts (nextjs / vercel) for getting the parsed geo data?

If It helps, here's my code app/ping/route.ts

```ts
import { NextRequest, NextResponse } from "next/server";

export function GET(req: NextRequest) {
  const geo = {
    country: req.geo?.country,
    latitude: req.geo?.latitude,
  };

  const headers: Record<string, string> = {};
  req.headers.forEach((value, key) => {
    headers[key] = value;
  });

  const searchparams: Record<string, string> = {};
  req.nextUrl.searchParams.forEach((value, key) => {
    searchparams[key] = value;
  });

  return NextResponse.json({
    geo,
    headers,
    searchparams,
  });
}
```

and this route will return below (some data was manually masked before taking screenshot)

<img width="426" alt="Screenshot 2024-08-31 at 3 19 22 AM" src="https://github.com/user-attachments/assets/906e43ad-98aa-4153-b8fe-9789616b639f">

### Is there any context that might help us understand?

nope.

### Does the docs page already exist? Please link to it.

https://nextjs.org/docs/app/api-reference/functions/next-request#geo

**Comment by samcx** [comment]
@softmarshmallow Thank you for noticing!

We are actually looking to remove this from req in Middleware, since this actually a platform feature (to have Next.js remain more platform agnostic). We recommend using the helper in `@vercel/functions` instead.
